src/notebooks/execution/cellExecutionMessageHandler.ts

```diff
@@ -0,0 +1,574 @@
+        error: Error;
+        msg: KernelMessage.IIOPubMessage;
+    }>();
+    public readonly onErrorHandlingExecuteRequestIOPubMessage = this._onErrorHandlingIOPubMessage.event;
+    private temporaryExecution?: NotebookCellExecution;
+    private previousResultsToRestore?: NotebookCellExecutionSummary;
+    private cellHasErrorsInOutput?: boolean;
+
+    public get hasErrorOutput() {
+        return this.cellHasErrorsInOutput === true;
+    }
+    /**
+     * We keep track of the last output that was used to store stream text.
+     * We need this so that we can update it later on (when we get new data for the same stream).
+     * If users clear outputs or if we have a new output other than stream, then clear this item.
+     * Because if after the stream we have an image, then the stream is not the last output item, hence its cleared.
+     */
+    private lastUsedStreamOutput?: { stream: 'stdout' | 'stderr'; text: string; output: NotebookCellOutput };
+    private readonly disposables: IDisposable[] = [];
+    private readonly prompts = new Set<CancellationTokenSource>();
+    constructor(
+        public readonly cell: NotebookCell,
+        private readonly applicationService: IApplicationShell,
+        private readonly controller: NotebookController,
+        private readonly outputDisplayIdTracker: CellOutputDisplayIdTracker,
+        private readonly context: IExtensionContext,
+        private readonly formatters: ITracebackFormatter[],
+        private readonly kernel: Kernel.IKernelConnection,
+        request: Kernel.IShellFuture<KernelMessage.IExecuteRequestMsg, KernelMessage.IExecuteReplyMsg>,
+        cellExecution: NotebookCellExecution
+    ) {
+        workspace.onDidChangeNotebookDocument(
+            (e) => {
+                if (!isJupyterNotebook(e.notebook)) {
+                    return;
+                }
+                const thisCellChange = e.cellChanges.find(({ cell }) => cell === this.cell);
+                if (!thisCellChange) {
+                    return;
+                }
+                if (thisCellChange.outputs?.length === 0) {
+                    // keep track of the fact that user has cleared the output.
+                    this.clearLastUsedStreamOutput();
+                    this.cellHasErrorsInOutput = false;
+                }
+            },
+            this,
+            this.disposables
+        );
+        this.execution = cellExecution;
+        this.startHandlingExecutionMessages(request);
+    }
+    /**
+     * This method is called when all execution has been completed (successfully or failed).
+     * Or when execution has been cancelled.
+     */
+    public dispose() {
+        traceCellMessage(this.cell, 'Execution disposed');
+        disposeAllDisposables(this.disposables);
+        this.prompts.forEach((item) => item.dispose());
+        this.prompts.clear();
+    }
+    public endCellExecution() {
+        this.prompts.forEach((item) => item.dispose());
```

Comment:
Wouldn't dispose call this function?

Reply:
that too yes, but when cell execution ends we don't want to dispose this straight away. That's more clear in the second PR, shoot, should have moved this to the next PR.

This is what we need:
* When we're executing a cell, this class will handle all of the outptu.
* Once execution completes, we need it to stop handling prompts and the like, hence we call endCellExecution
* But we don't want to dispose it just yet (in PR 3) - we stop disposing it
* Sorry, I shoulld have created the method endCellExecution in PR 3
* In PR 3, what we do is when execution ends, we call endCelLExecution
* However that class lives on and keeps monitoring the  iopub messages
* Dispose is called when the cell is deleted.